weave/tests/trace/test_evaluation_performance.py

```diff
@@ -0,0 +1,165 @@
+        if (
+            item == "internal_trace_server"
+            or item == "lock"
+            or item == "resume"
+            or item == "pause"
+        ):
+            return super().__getattribute__(item)
+        internal_trace_server = super().__getattribute__("internal_trace_server")
+
+        if item == "attribute_access_log":
+            return getattr(internal_trace_server, item)
+
+        def wrapper(*args, **kwargs):
+            with self.lock:
+                inner_attr = getattr(internal_trace_server, item)
+                result = inner_attr(*args, **kwargs)
+            return result
+
+        return wrapper
+
+
+@contextmanager
+def paused_client(client: WeaveClient) -> Generator[WeaveClient, None, None]:
+    original_server = client.server
+    client.set_autoflush(False)
+    blocking_server = BlockingTraceServer(original_server)
+    client.server = blocking_server
+    blocking_server.pause()
+    try:
+        yield client
+    finally:
+        blocking_server.resume()
+        client.server = original_server
+        client._flush()
+        client.set_autoflush(True)
+
+
+def build_evaluation():
+    dataset = [
+        {
+            "question": "What is the capital of France?",
+            "expected": "Paris",
+            "img": PIL.Image.new("RGB", (100, 100)),
+        },
+        {
+            "question": "Who wrote 'To Kill a Mockingbird'?",
+            "expected": "Harper Lee",
+            "img": PIL.Image.new("RGB", (100, 100)),
+        },
+        {
+            "question": "What is the square root of 64?",
+            "expected": "8",
+            "img": PIL.Image.new("RGB", (100, 100)),
+        },
+        {
+            "question": "What is the thing you say when you don't know something?",
+            "expected": "I don't know",
+            "img": PIL.Image.new("RGB", (100, 100)),
+        },
+    ]
+
+    @weave.op()
+    def predict(question: str):
+        return "I don't know"
+
+    @weave.op()
+    def score(question: str, expected: str, model_output: str):
+        return model_output == expected
+
+    evaluation = weave.Evaluation(
+        name="My Evaluation",
+        dataset=dataset,
+        scorers=[score],
+    )
+
+    return evaluation, predict
+
+
+@pytest.mark.asyncio
+async def test_evaluation_performance(client: WeaveClient):
+    client.project = "test_evaluation_performance"
+    evaluation, predict = build_evaluation()
+
+    log = [l for l in client.server.attribute_access_log if not l.startswith("_")]
+
+    assert log == ["ensure_project_exists"]
+
+    with paused_client(client) as client:
+        res = await evaluation.evaluate(predict)
+        assert res["score"]["true_count"] == 1
+        log = [l for l in client.server.attribute_access_log if not l.startswith("_")]
+        assert log == ["ensure_project_exists"]
+
+    log = [l for l in client.server.attribute_access_log if not l.startswith("_")]
```

Comment:
Can you help me grok this?

It seems like:
1. You have a test checking for the internal sever calls, where the first is the `ensure_project_exists` check
2. There is the same thing(?) with a `paused_client` context?  What is this for?